### PR TITLE
Fix the pipeline breaks dues to the MSVC 19.40 and numpy 2.0 release

### PR DESCRIPTION
**Pull request by wenbingl**
inspired by this PR: https://github.com/microsoft/onnxruntime/pull/21004

### CODE REVIEW

**Comment by wenbingl** [comment]
@snnn any idea why the following command line doesn't work


2024-06-14T23:22:52.1858496Z   "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -S D:\a\_work\1\s -B .scb\temp.win-amd64-cpython-312\Release -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=D:\a\_work\1\s\.scb\lib.win-amd64-cpython-312\onnxruntime_extensions -DOCOS_ENABLE_CTEST=OFF -DOCOS_BUILD_PYTHON=ON -DOCOS_PYTHON_MODULE_PATH=D:\a\_work\1\s\.scb\lib.win-amd64-cpython-312\onnxruntime_extensions\_extensions_pydll.cp312-win_amd64.pyd **-DCMAKE_BUILD_TYPE=Release -DOCOS_USE_CUDA=ON "-T cuda=D:\a\_work\_temp\v11.8" -DCMAKE_CUDA_FLAGS_INIT=--allow-unsupported-compiler -DCMAKE_CUDA_ARCHITECTURES=70;86**

**Comment by snnn** [comment]
Use the latest cmake

**Comment by snnn** [comment]
Remove "python -m pip install cmake"

**Comment by snnn** [comment]
The machine already has cmake

File: .pyproject/cmdclass.py [hunk 1]
```diff
@@ -226,6 +226,8 @@ def build_cmake(self, extension):
                 if sys.platform == "win32":
                     cuda_path = os.environ.get("CUDA_PATH")
                     cmake_args += [f'-T cuda={cuda_path}']
+                    # TODO: temporarily add a flag for MSVC 19.40
+                    cmake_args += ['-DCMAKE_CUDA_FLAGS_INIT=--allow-unsupported-compiler']
```
Comment (snnn) [inline .pyproject/cmdclass.py:230]: Only one dash

File: cmake/externals/opencv-no-rtti.patch [hunk 2]
```diff
@@ -39,6 +39,19 @@ index d95e5db163..db185453df 100644
  include(cmake/OpenCVCompilerOptions.cmake)
  
  ocv_cmake_hook(POST_COMPILER_OPTIONS)
+diff --git a/cmake/OpenCVDetectCXXCompiler.cmake b/cmake/OpenCVDetectCXXCompiler.cmake
+index 7f229cde96..92e204a5b9 100644
+--- a/cmake/OpenCVDetectCXXCompiler.cmake
++++ b/cmake/OpenCVDetectCXXCompiler.cmake
+@@ -171,7 +171,7 @@ elseif(MSVC)
```
Comment (snnn) [inline cmake/externals/opencv-no-rtti.patch:46]: @yuslepukhin FYI

File: .pipelines/ci.yml [hunk 3]
```diff
@@ -570,7 +570,9 @@ stages:
 
       - script: |
           set CUDA_PATH=$(Agent.TempDirectory)\v11.8
-          call .\build.bat -T cuda="%CUDA_PATH%" -DOCOS_ENABLE_CTEST=ON -DOCOS_USE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=70;86^
+          call .\build.bat -T cuda="%CUDA_PATH%" -DOCOS_ENABLE_CTEST=ON^
+            -DCMAKE_CUDA_FLAGS_INIT=--allow-unsupported-compiler^
```
Comment (yuslepukhin) [inline .pipelines/ci.yml:574]: One dash

Comment (wenbingl) [reply .pipelines/ci.yml:574]: the initial commit is one dash which failed too, and some examples from GH suggest two dashes, but it doesn't work either. 
According to the log, CMAKE_CUDA_FLAGS_INIT doesn't change the nvcc command line, need to figure that out.

**Comment by snnn** [comment]
Do not use `C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe`
Use 
```
 C:\Program Files\CMake\bin\cmake.exe
```

**Comment by snnn** [comment]
I deleted one of the cmake from the VM image. Then one of the failed pipelines passed.

File: CMakeLists.txt [hunk 4]
```diff
@@ -36,6 +36,11 @@ if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.24.0")
   cmake_policy(SET CMP0077 NEW)
 endif()
 
+# Avoid warning of Calling FetchContent_Populate(GSL) is deprecated
```
Comment (snnn) [inline CMakeLists.txt:39]: Thanks for the fix. How did you find it? CMake 3.30 is not release yet.

**Comment by wenbingl** [comment]
> I deleted one of the cmake from the VM image. Then one of the failed pipelines passed.

Why is there no C:\program files\cmake\... in CUDA Windows machine?

File: CMakeLists.txt (hunk 4 above)
Comment (wenbingl) [reply CMakeLists.txt:39]: I tried several versions of cmake local installer and noticed this coming change.

**Review by sayanshaw24 — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Comment by snnn** [comment]
> > I deleted one of the cmake from the VM image. Then one of the failed pipelines passed.
> 
> Why is there no C:\program files\cmake... in CUDA Windows machine?

It shouldn't. Could you give me a link of the log?

**Comment by wenbingl** [comment]
> > > I deleted one of the cmake from the VM image. Then one of the failed pipelines passed.
> > 
> > 
> > Why is there no C:\program files\cmake... in CUDA Windows machine?
> 
> It shouldn't. Could you give me a link of the log?

There is no direct log. Python code detected ProgramFiles/CMake firstly then falled back to VS/cmake. In this log, https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1409852&view=logs&j=1130dc32-b83a-5b28-25cb-8e5ca42e48d6&t=524fdcc1-2f40-5737-3487-03bf6009139b, it did fallback to vs/cmake.

**Comment by woinck** [comment]
Hi @wenbingl , CMake 3.30 is failing with this PR due to this line: https://github.com/microsoft/onnxruntime-extensions/blob/3b275b16bc7a11fb89f82546d1209a161ab057ca/CMakeLists.txt#L41
I think it's because it's missing SET argument in cmake_policy. Is it possible to fix it? Thanks!

**Comment by wenbingl** [comment]
> Hi @wenbingl , CMake 3.30 is failing with this PR due to this line:
> 
> https://github.com/microsoft/onnxruntime-extensions/blob/3b275b16bc7a11fb89f82546d1209a161ab057ca/CMakeLists.txt#L41
> 
> 
> I think it's because it's missing SET argument in cmake_policy. Is it possible to fix it? Thanks!

Thanks for the reporting, here is the fixing: https://github.com/microsoft/onnxruntime-extensions/pull/754

**Pull request merged**